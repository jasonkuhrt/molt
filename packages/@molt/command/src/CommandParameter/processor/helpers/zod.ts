import type { Pam } from '../../../lib/Pam/index.js'
import type { ZodNumberCheck, ZodStringCheck } from '../../../lib/zodHelpers/index_.js'
import type { SomeBasicType, SomeUnionType, TypeNumber, TypeString } from '../../types.js'
import { Alge } from 'alge'
import { z } from 'zod'

// export const analyzeZodType = (zodType: SomeBasicType | SomeUnionType) => {
//   if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodUnion) {
//     let description = zodType.description ?? null
//     let primitiveType = zodType

//     while (
//       primitiveType._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault ||
//       primitiveType._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional
//     ) {
//       description = description ?? primitiveType._def.innerType.description ?? null
//       primitiveType = primitiveType._def.innerType
//     }

//     const members = zodType._def.options.map((_) => {
//       const typeAnalysis = analyzeZodTypeScalar(_)
//       return {
//         // zodType: _,
//         description: typeAnalysis.description,
//         type: typeAnalysis.type,
//       }
//     })
//     return {
//       _tag: `TypeUnion`,
//       members,
//     }
//   }

//   return analyzeZodTypeScalar(zodType)
// }

export const analyzeZodType = (zodType: SomeBasicType | SomeUnionType) => {
  return analyzeZodType_(zodType, zodType.description ?? null)
}

// prettier-ignore
type ZodTypeToType<ZT extends SomeBasicType | SomeUnionType> =
  ZT extends z.ZodOptional<infer T>       ? ZodTypeToType<T> :
  ZT extends z.ZodDefault<infer T>        ? ZodTypeToType<T> :
  ZT extends z.ZodLiteral<any>            ? Pam.Type.Literal :
  ZT extends z.ZodString                  ? Pam.Type.String :
  ZT extends z.ZodBoolean                 ? Pam.Type.Boolean :
  ZT extends z.ZodNumber                  ? Pam.Type.Number :
  ZT extends z.ZodEnum<any>               ? Pam.Type.Enumeration :
  ZT extends z.ZodNativeEnum<any>         ? Pam.Type.Enumeration :
  ZT extends z.ZodUnion<any>              ? Pam.Type.Union :
                                            never

const analyzeZodType_ = <ZT extends SomeBasicType | SomeUnionType>(
  zodType: ZT,
  description: null | string,
): {
  type: ZodTypeToType<ZT>
  // type: ZT
  description: null | string
} => {
  if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault) {
    return analyzeZodType_(zodType._def.innerType as ZT, description)
  }

  if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional) {
    return analyzeZodType_(zodType._def.innerType as ZT, description)
  }

  const type: null | Pam.Type.Group.Any =
    zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodUnion
      ? {
          _tag: `TypeUnion` as const,
          members: zodType._def.options.map((_) => {
            const typeAnalysis = analyzeZodType(_)
            return {
              // zodType: _,
              description: typeAnalysis.description,
              type: typeAnalysis.type,
            }
          }),
        }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodLiteral
      ? { _tag: `TypeLiteral`, value: zodType._def.value }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodString
      ? { _tag: `TypeString`, ...mapZodStringChecks(zodType._def.checks) }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodBoolean
      ? { _tag: `TypeBoolean` }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodNumber
      ? { _tag: `TypeNumber`, ...mapZodNumberChecks(zodType._def.checks) }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodNativeEnum
      ? { _tag: `TypeEnum`, members: Object.values(zodType._def.values) }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodEnum
      ? { _tag: `TypeEnum`, members: zodType._def.values }
      : null

  if (!type) throw new Error(`Unsupported zod type: ${zodType._def.typeName}`)

  return {
    description,
    type: type as ZodTypeToType<ZT>,
  }
}

const mapZodNumberChecks = (checks: z.ZodNumberCheck[]) => {
  return checks
    .map((_): ZodNumberCheck => ({ _tag: _.kind, ..._ }) as any)
    .reduce(
      (acc, check) => {
        return Alge.match(check)
          .int(() => ({ ...acc, int: true }))
          .min((check) => ({
            min: check.value,
          }))
          .max((check) => ({
            max: check.value,
          }))
          .finite(() => ({
            finite: true,
          }))
          .multipleOf((check) => ({
            multipleOf: check.value,
          }))
          .done()
      },
      {} as Omit<TypeNumber, '_tag'>,
    )
}

const mapZodStringChecks = (checks: z.ZodStringCheck[]): Omit<TypeString, '_tag'> => {
  return checks
    .map((_): ZodStringCheck => ({ _tag: _.kind, ..._ }) as any)
    .reduce<Omit<TypeString, '_tag'>>(
      (acc, check) => {
        return {
          ...acc,
          ...Alge.match(check)
            .regex((check) => ({
              regex: check.regex,
            }))
            .min((check) => ({
              min: check.value,
            }))
            .max((check) => ({
              max: check.value,
            }))
            .url((check) => ({
              pattern: { type: check._tag },
            }))
            .cuid((check) => ({
              pattern: { type: check._tag },
            }))
            .cuid2(() => ({
              pattern: { type: `cuid2` as const },
            }))
            .uuid((check) => ({
              pattern: { type: check._tag },
            }))
            .emoji((_) => ({
              pattern: { type: _._tag },
            }))
            .ip((_) => ({
              pattern: {
                type: _._tag,
                version: _.version
                  ? Alge.match(_.version)
                      .v4(() => 4 as const)
                      .v6(() => 6 as const)
                      .done()
                  : null,
              },
            }))
            .ulid((_) => ({
              pattern: { type: _._tag },
            }))
            .datetime((check) => ({
              pattern: {
                type: `dateTime` as const,
                offset: check.offset,
                precision: check.precision,
              },
            }))
            .email((check) => ({
              pattern: { type: check._tag },
            }))
            .endsWith((check) => ({
              endsWith: check.value,
            }))
            .startsWith((check) => ({
              startsWith: check.value,
            }))
            .length((check) => ({
              length: check.value,
            }))
            .includes((_) => ({
              includes: _.value,
            }))
            // transformations
            .trim(() => ({
              transformations: {
                ...acc.transformations,
                trim: true,
              },
            }))
            .toLowerCase(() => ({
              transformations: {
                ...acc.transformations,
                toCase: `lower` as const,
              },
            }))
            .toUpperCase(() => ({
              transformations: {
                ...acc.transformations,
                toCase: `upper` as const,
              },
            }))
            .done(),
        }
      },
      {} as Omit<TypeString, '_tag'>,
    )
}
