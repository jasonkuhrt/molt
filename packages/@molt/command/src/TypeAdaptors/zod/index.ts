export * as Zod from './zod.js'
