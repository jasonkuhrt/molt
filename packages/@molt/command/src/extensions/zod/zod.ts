import type { CommandParameter } from '../../CommandParameter/index.js'
import { createExtension } from '../../extension.js'
import type { HKT } from '../../helpers.js'
import { TypeAdaptor } from './TypeAdaptor/index.js'
import type { z } from 'zod'

export type ZodType = CommandParameter.SomeBasicType | CommandParameter.SomeUnionType

interface ZodTypeMapper extends HKT.Fn<z.ZodTypeAny> {
  return: TypeAdaptor.Zod.FromZod<this['params']>
}

export const Zod = createExtension<ZodType, ZodTypeMapper>({
  name: `Zod`,
  type: (zodType: z.ZodFirstPartySchemaTypes) => TypeAdaptor.Zod.fromZod(zodType) as any,
})
