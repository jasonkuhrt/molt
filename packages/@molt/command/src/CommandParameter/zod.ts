import { ZodHelpers } from '../lib/zodHelpers/index.js'
import { stripOptionalAndDefault } from '../lib/zodHelpers/index_.js'
import { z } from 'zod'

export type SomeBasicParameterType = SomeBasicType | SomeUnionType

type ZodEnumBase = z.ZodEnum<[string, ...string[]]>

type ZodNativeEnumBase = z.ZodNativeEnum<any>

export type SomeBasicType =
  | SomeBasicTypeScalar
  | z.ZodOptional<SomeBasicTypeScalar>
  | z.ZodDefault<SomeBasicTypeScalar>

export type SomeUnionType = SomeUnionTypeScalar | z.ZodOptional<SomeUnionType> | z.ZodDefault<SomeUnionType>

export type SomeExclusiveZodType = SomeBasicTypeScalar

export type SomeUnionTypeScalar = z.ZodUnion<
  [SomeBasicTypeScalar, SomeBasicTypeScalar, ...SomeBasicTypeScalar[]]
>

// prettier-ignore
export type SomeBasicTypeScalar =
  | z.ZodString
  | ZodEnumBase
  | ZodNativeEnumBase
  | z.ZodNumber
  | z.ZodBoolean
  | z.ZodLiteral<number>
  | z.ZodLiteral<string>
  | z.ZodLiteral<boolean>

export const isUnionType = (type: SomeBasicType | SomeUnionType): type is SomeUnionType => {
  const type_ = stripOptionalAndDefault(type)
  const isUnion = type_._def.typeName === z.ZodFirstPartyTypeKind.ZodUnion
  return isUnion
}

export const getUnionScalar = (zodType: SomeUnionType): SomeUnionTypeScalar => {
  const zodScalarType = ZodHelpers.stripOptionalAndDefault(zodType)
  return zodScalarType
}

export const getBasicScalar = (zodType: SomeBasicType): SomeBasicTypeScalar => {
  const type = ZodHelpers.stripOptionalAndDefault(zodType)
  return type
}
