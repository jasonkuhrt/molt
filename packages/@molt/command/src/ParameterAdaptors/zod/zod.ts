import { TypeAdaptors } from '../../TypeAdaptors/index.js'
import { z } from 'zod'

export const inferPropsFromZodType = (
  zodType: TypeAdaptors.Zod.SomeBasicType | TypeAdaptors.Zod.SomeUnionType,
) => {
  const isOptional = zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional
  const hasDefault = zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault
  const defaultGetter = hasDefault ? (zodType._def.defaultValue as DefaultGetter) : null
  const optionality: Pam.Optionality = defaultGetter
    ? { _tag: `default`, getValue: () => defaultGetter() }
    : isOptional
    ? { _tag: `optional` }
    : { _tag: `required` }

  const { type, description } = TypeAdaptors.Zod.analyzeZodType(zodType)

  return {
    optionality,
    description,
    // Cannot use @ts-expect-error because someone the build then
    // does _not_ detect an error here, in which case the expect error
    // itself triggers the error... chicken and egg problem.
    //
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore todo
    type: type,
  }
}
