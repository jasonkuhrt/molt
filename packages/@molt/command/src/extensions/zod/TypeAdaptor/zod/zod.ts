import { Type } from '../../../../Type/index.js'
import { ZodHelpers } from '../../zodHelpers/index.js'
import { Alge } from 'alge'
import type { Simplify } from 'type-fest'
import { z } from 'zod'

// prettier-ignore
export type FromZod<ZodType extends z.ZodType> =
  ZodType extends z.ZodOptional<infer T>
    ? Type.Union<[FromZodNonOptional<T>, Type.Literal<undefined>]>
    : FromZodNonOptional<ZodType>

// prettier-ignore
export type FromZodNonOptional<ZodType extends z.ZodType> =
  ZodType extends ZodTypeScalar              ? FromZodScalar<ZodType> :
  ZodType extends z.ZodDefault<infer T>      ? FromZodScalar<T> :
  ZodType extends z.ZodUnion<infer T>        ? Type.Union<{[i in keyof T]: FromZodScalar<T[i]>}> :
                                                never

// prettier-ignore
export type FromZodScalar<ZodType extends ZodTypeScalar> =
  ZodType extends z.ZodString                                           ? Type.Scalar.String :
  ZodType extends z.ZodBoolean                                          ? Type.Scalar.Boolean :
  ZodType extends z.ZodLiteral<infer T extends boolean|string|number>   ? Type.Scalar.Literal<T> :
  ZodType extends z.ZodNumber                                           ? Type.Scalar.Number :
  ZodType extends z.ZodEnum<infer T>                                    ? Type.Scalar.Enumeration<T> :
  ZodType extends z.ZodNativeEnum<infer T extends z.EnumLike>           ? Type.Scalar.Enumeration<EnumerationMembersFromZodEnumLike<T>> :
                                                                          never
type EnumerationMembersFromZodEnumLike<T extends z.EnumLike> = T[keyof T][]

type ZodTypeScalar =
  | z.ZodString
  | z.ZodBoolean
  | z.ZodLiteral<any>
  | z.ZodNumber
  | z.ZodEnum<any>
  | z.ZodNativeEnum<any>
// prettier-ignore

export const fromZod = (zodType: z.ZodFirstPartySchemaTypes): Type.Type => _fromZod(zodType, {})

// prettier-ignore
const _fromZod = (zodType: z.ZodFirstPartySchemaTypes,previous:{description?:string;optionality?:Type.Optionality<any>}): Type.Type => {
  const zt = zodType
  const description = previous.description ?? zt.description

  const isOptional = zt._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional
  const hasDefault = zt._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault
  const defaultGetter = hasDefault ? (zt._def.defaultValue as () => unknown) : null
  const optionality: Type.Optionality<any> = previous.optionality ?? (defaultGetter
    ? { _tag: `default`, getValue: () => defaultGetter() }
    : isOptional
    ? { _tag: `optional` }
    : { _tag: `required` })
  
  if (ZodHelpers.isString(zt)) {
    const {refinements,transformations} = mapZodStringChecksAndTransformations(zt._def.checks)
    return Type.string({ optionality, refinements, transformations, description })
  }
  if (ZodHelpers.isNumber(zt)) {
    const {refinements} = mapZodNumberChecksAndTransformations(zt._def.checks)
    return Type.number({ optionality, refinements, description })
  }
  if (ZodHelpers.isEnum(zt))            return Type.enumeration({ optionality, members: zt._def.values, description }) // eslint-disable-line
  if (ZodHelpers.isNativeEnum(zt))      return Type.enumeration({ optionality, members: Object.values<any>(zt._def.values), description })
  if (ZodHelpers.isBoolean(zt))         return Type.boolean({ optionality, description })
  if (ZodHelpers.isLiteral(zt))         return Type.literal({ optionality, value: zt._def.value, description }) // eslint-disable-line
  if (ZodHelpers.isDefault(zt))         return _fromZod(zt._def.innerType,{...previous,description,optionality})
  if (ZodHelpers.isOptional(zt))        return _fromZod(zt._def.innerType,{...previous,description,optionality})
  if (ZodHelpers.isUnion(zt))           {
    if (!Array.isArray(zt._def.options)) throw new Error(`Unsupported zodType: ${JSON.stringify(zt[`_def`])}`)
    const members = zt._def.options.map((_: {_def:{description:undefined|string}}) => {
      const description = _._def.description
      return _fromZod(_ as any, {...previous,description,optionality})
    })
    return Type.union({ optionality, members_: members, description })
  }
  throw new Error(`Unsupported zodType: ${JSON.stringify(zt[`_def`])}`)
}

const mapZodNumberChecksAndTransformations = (checks: z.ZodNumberCheck[]) => {
  return checks.reduce(
    (acc, check) => {
      return {
        refinements: {
          ...acc.refinements,
          ...Alge.match(check)
            .int(() => ({ ...acc, int: true }))
            .min((check) => ({
              min: check.value,
            }))
            .max((check) => ({
              max: check.value,
            }))
            .finite(() => ({
              finite: true,
            }))
            .multipleOf((check) => ({
              multipleOf: check.value,
            }))
            .done(),
        },
      }
    },
    {} as Simplify<Pick<Type.Scalar.Number, 'refinements'>>,
  )
}

const mapZodStringChecksAndTransformations = (checks: z.ZodStringCheck[]) => {
  const transformations = [`trim`, `toCase`] as const
  return checks.reduce(
    (acc, check) => {
      const refinementOrTransformation = Alge.match(check)
        .regex((check) => ({
          regex: check.regex,
        }))
        .min((check) => ({
          min: check.value,
        }))
        .max((check) => ({
          max: check.value,
        }))
        .url((check) => ({
          pattern: { type: check.kind },
        }))
        .cuid((check) => ({
          pattern: { type: check.kind },
        }))
        .cuid2(() => ({
          pattern: { type: `cuid2` as const },
        }))
        .uuid((check) => ({
          pattern: { type: check.kind },
        }))
        .emoji((_) => ({
          pattern: { type: _.kind },
        }))
        .ip((_) => ({
          pattern: {
            type: _.kind,
            version: _.version
              ? Alge.match(_.version)
                  .v4(() => 4 as const)
                  .v6(() => 6 as const)
                  .done()
              : null,
          },
        }))
        .ulid((_) => ({
          pattern: { type: _.kind },
        }))
        .datetime((check) => ({
          pattern: {
            type: `dateTime` as const,
            offset: check.offset,
            precision: check.precision,
          },
        }))
        .email((check) => ({
          pattern: { type: check.kind },
        }))
        .endsWith((check) => ({
          endsWith: check.value,
        }))
        .startsWith((check) => ({
          startsWith: check.value,
        }))
        .length((check) => ({
          length: check.value,
        }))
        .includes((_) => ({
          includes: _.value,
        }))
        // transformations
        .trim(() => ({
          trim: true,
        }))
        .toLowerCase(() => ({
          toCase: `lower` as const,
        }))
        .toUpperCase(() => ({
          toCase: `upper` as const,
        }))
        .done()

      const isTransformation = transformations.includes(Object.keys(refinementOrTransformation)[0]! as any)
      return isTransformation
        ? {
            ...acc,
            transformations: {
              ...acc.transformations,
              ...refinementOrTransformation,
            },
          }
        : {
            ...acc,
            refinements: {
              ...acc.refinements,
              ...refinementOrTransformation,
            },
          }
    },
    {} as Simplify<Pick<Type.Scalar.String, 'refinements' | 'transformations'>>,
  )
}
