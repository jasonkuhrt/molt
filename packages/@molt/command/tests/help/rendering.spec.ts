import { Command } from '../../src/index.js'
import stripAnsi from 'strip-ansi'
import { describe, expect, it } from 'vitest'
import { mockProcessExit, mockProcessStdout } from 'vitest-mock-process'
import { z } from 'zod'

const processStdout = mockProcessStdout()
mockProcessExit()

it(`if there is optional param it is shown`, () => {
  Command.create({ a: z.string().optional() }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`if parameter has description it is shown`, () => {
  Command.create({ a: z.string().optional().describe(`Blah blah blah.`) }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`long description wraps within column`, () => {
  Command.create({
    a: z.string().optional().describe(`Blah blah blah. Blah blah blah. Blah blah blah.`),
  }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`if parameter has default it is shown`, () => {
  Command.create({ foo: z.string().default(`bar`) }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`if parameter is optional without default then its default shows up as "undefined"`, () => {
  Command.create({ foo: z.string().optional() }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`if there is an error trying to get default then a nice message is shown`, () => {
  Command.create({
    foo: z.string().default(() => {
      throw new Error(`whoops`)
    }),
  }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`if there is an error trying to get default then a nice message is shown`, () => {
  Command.create({
    foo: z.string().default(() => {
      throw new Error(`whoops`)
    }),
  }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

it(`enums do not mess up alignment when they are widest line in the column`, () => {
  // prettier-ignore
  Command.create({
    foo: z.enum([`a`, `b`, `c`, `d`, `e`, `f`, `g`, `h`, `i`, `j`, `k`, `l`, `m`, `n`, `o`, `p`, `q`, `r`, `s`, `t`, `u`, `v`, `w`, `x`, `y`, `z`]),
    bar: z.string().optional(),
  }).parseOrThrow([`-h`])
  const output = processStdout.mock.lastCall?.[0] as string
  expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
  expect(output).toMatchSnapshot(`polychrome`)
})

describe(`enum`, () => {
  it(`enum members are listed`, () => {
    Command.create({ foo: z.enum([`apple`, `dolphin`, `cab`]) }).parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })

  it(`optional enum members are listed`, () => {
    Command.create({ foo: z.enum([`apple`, `dolphin`, `cab`]).optional() }).parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })

  it(`when there is only one enum member it is prefixed with "enum:" to avoid confusion of it looking like the name of a kind of type`, () => {
    Command.create({ foo: z.enum([`apple`]) }).parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })

  it(`when there are many members they wrap`, () => {
    Command.create({
      foo: z.enum([
        `apple`,
        `baby`,
        `cannabis`,
        `dinosaur`,
        `elephant`,
        `fanna`,
        `goat`,
        `house`,
        `island`,
        `jake`,
        `kilomanjara`,
      ]),
    }).parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
})

describe(`environment`, () => {
  it(`when environment is disabled then environment doc is not shown`, () => {
    Command.create({ foo: z.string() })
      .settings({ parameters: { environment: false } })
      .parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
  it(`when environment is enabled it shows as the last column`, () => {
    Command.create({ foo: z.string() })
      .settings({ parameters: { environment: true } })
      .parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
  it(`when environment is disabled for one parameter it has X indicating that`, () => {
    Command.create({ foo: z.string(), bar: z.string() })
      .settings({ parameters: { environment: { $default: true, foo: false } } })
      .parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
  it(`when environment has custom prefix it is displayed`, () => {
    Command.create({ foo: z.string(), bar: z.string() })
      .settings({ parameters: { environment: { $default: true, foo: { prefix: `moo` } } } })
      .parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
  it(`when environment has multiple custom prefix they are displayed`, () => {
    Command.create({ foo: z.string(), bar: z.string() })
      .settings({ parameters: { environment: { $default: true, foo: { prefix: [`moo`, `boo`] } } } })
      .parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
  it(`when environment has no prefix it is displayed`, () => {
    Command.create({ foo: z.string(), bar: z.string() })
      .settings({ parameters: { environment: { $default: true, foo: { prefix: false } } } })
      .parseOrThrow([`-h`])
    const output = processStdout.mock.lastCall?.[0] as string
    expect(stripAnsi(output)).toMatchSnapshot(`monochrome`)
    expect(output).toMatchSnapshot(`polychrome`)
  })
})
