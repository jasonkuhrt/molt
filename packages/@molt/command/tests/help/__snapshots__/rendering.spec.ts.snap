// Vitest Snapshot v1

exports[`enum members are listed > monochrome 1`] = `
"
PARAMETERS

  
  foo    apple | dolphin | cab    REQUIRED            
  
"
`;

exports[`enum members are listed > polychrome 1`] = `
"
PARAMETERS

  
  [32mfoo[39m    [32mapple[39m[33m | [39m[32mdolphin[39m[33m | [39m[32mcab[39m   [101m REQUIRED [49m           
  
"
`;

exports[`if parameter has default it is shown > monochrome 1`] = `
"
PARAMETERS

  
  a      string    
  a                
  
"
`;

exports[`if parameter has default it is shown > polychrome 1`] = `
"
PARAMETERS

  
  [32ma[39m      [32mstring[39m    
  [90ma[39m                
  
"
`;

exports[`if parameter has description it is shown > monochrome 1`] = `
"
PARAMETERS

  
  a      string            
  a      Blah blah blah.   
  
"
`;

exports[`if parameter has description it is shown > polychrome 1`] = `
"
PARAMETERS

  
  [32ma[39m      [32mstring[39m            
  [90ma[39m      Blah blah blah.   
  
"
`;

exports[`if there is optional param it is shown > monochrome 1`] = `
"
PARAMETERS

  
  a      string    
  a                
  
"
`;

exports[`if there is optional param it is shown > polychrome 1`] = `
"
PARAMETERS

  
  [32ma[39m      [32mstring[39m    
  [90ma[39m                
  
"
`;

exports[`long description wraps within column > monochrome 1`] = `
"
PARAMETERS

  
  a      string                                     
  a      Blah blah blah. Blah blah blah. Blah bla   
         h blah.                                    
  
"
`;

exports[`long description wraps within column > polychrome 1`] = `
"
PARAMETERS

  
  [32ma[39m      [32mstring[39m                                     
  [90ma[39m      Blah blah blah. Blah blah blah. Blah bla   
         h blah.                                    
  
"
`;

exports[`optional enum members are listed > monochrome 1`] = `
"
PARAMETERS

  
  foo    apple | dolphin | cab   
  
"
`;

exports[`optional enum members are listed > polychrome 1`] = `
"
PARAMETERS

  
  [32mfoo[39m    [32mapple[39m[33m | [39m[32mdolphin[39m[33m | [39m[32mcab[39m   
  
"
`;

exports[`when there are many members they wrap > monochrome 1`] = `
"
PARAMETERS

  
  foo    apple | baby | cannabis | dinosaur    REQUIRED            
         fanna | goat | house | island | jake                        
  
"
`;

exports[`when there are many members they wrap > polychrome 1`] = `
"
PARAMETERS

  
  [32mfoo[39m    [32mapple[39m[33m | [39m[32mbaby[39m[33m | [39m[32mcannabis[39m[33m | [39m[32mdinosaur[39m   [101m REQUIRED [49m           
         [32mfanna[39m[33m | [39m[32mgoat[39m[33m | [39m[32mhouse[39m[33m | [39m[32misland[39m[33m | [39m[32mjake[39m                        
  
"
`;

exports[`when there is only one enum member it is prefixed with "enum:" to avoid confusion of it looking like the name of a kind of type > monochrome 1`] = `
"
PARAMETERS

  
  foo    apple (enum)    REQUIRED            
  
"
`;

exports[`when there is only one enum member it is prefixed with "enum:" to avoid confusion of it looking like the name of a kind of type > polychrome 1`] = `
"
PARAMETERS

  
  [32mfoo[39m    [32mapple[39m [90m(enum)[39m   [101m REQUIRED [49m           
  
"
`;
